packages/flutter/lib/src/material/material_state_mixin.dart

```diff
@@ -63,7 +63,20 @@ mixin MaterialStateMixin<T extends StatefulWidget> on State<T> {
   /// To check for a single condition, convenience getters [isPressed], [isHovered],
   /// [isFocused], etc, are available for each [MaterialState] value.
   @protected
-  Set<MaterialState> materialStates = <MaterialState>{};
+  final WidgetStatesController widgetStatesController = WidgetStatesController();
```

Comment:
Having a WidgetStatesController inside the MaterialStateMixin seems wrong to me.

`MaterialStateMixin` (which I wasn't aware until today :smile: ) and `WidgetStatesController` (previously named `MaterialStatesController`) are two similar concepts.
Looking at the history they were both discussed in https://docs.google.com/document/d/1PJdsOoXhWdhnpYKlW5Nm2Di3Di5XARoYyHerMwWovXE/edit?tab=t.0

`MaterialStateMixin` was introduced in https://github.com/flutter/flutter/pull/82843 and did not get that much traction is it used only for Chip and RawMaterialButton (which is used for FAB and the deprecated MaterialButton).

`MaterialStatesController` was introduced in https://github.com/flutter/flutter/pull/103167/ . It is the preferred way to manage states.

I think removing Chip dependency on `MaterialStateMixin` is probably the way forward. The nice think with WidgetStatesController is that widget can accept an instance of it which gives users a way modify states and react to their changes. That's why several widgets accept such a parameter (and creates an internal controller if none is provided).

@nate-thegrate For insights and considering the deprecation of MaterialStateMixin.

Reply:
Yeah the issue I have here is that `Chip` is `MaterialStateMixin` but `InkWell` is using `WidgetStatesController`.

I don't mind changing `Chip`s to use `WidgetStatesController`. Should I include everything in this PR or should I try to split it in:

1. This PR doesn't include changes on `MaterialStateMixin` to only rely on `InkWell`'s current implementation (and it won't support `MaterialState.selected`)
2. 2nd PR to make `Chip` use `WidgetStatesController` instead of `MaterialStateMixin` (maybe we'd need to create a new issue for that part)
What do you think?

Reply:
If you can make step 1. without the changes on MaterialStateMixin, that's perfect.

For Step2, just found that there is already an issue, see https://github.com/flutter/flutter/issues/128289

Reply:
My opinion about MaterialStateMixin vs. WidgetStatesController is that both are nice, but the controller is a bit more intuitive, since it's similar to other Listenable objects in the framework.

That being said, I wanted to use a listenable object to control a set of WidgetStates in a personal project, and eventually I settled on a third-party pub package as my overall favorite:

<table align="center">
<tr>
<th>WidgetStatesController</th>
<th>SetNotifier&lt;WidgetState&gt;</th>
</tr>
<tr><td>

```dart
// add a state
states.update(WidgetState.error, true);

// remove a state
states.update(WidgetState.error, false);       
``` 

</td>
<td>

```dart
// add a state
states.add(WidgetState.error);

// remove a state
states.remove(WidgetState.error);       
```

</td></tr>
</table>

I really like how [**collection_notifiers**](https://pub.dev/packages/collection_notifiers) can turn a List/Set/Map into a listenable without really changing the API surface. Perhaps we could port it into the framework at some point.